presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.cpp

```diff
@@ -71,45 +95,93 @@ class PartitionAndSerializeOperator : public Operator {
 
   void addInput(RowVectorPtr input) override {
     input_ = std::move(input);
-  }
-
-  RowVectorPtr getOutput() override {
-    if (!input_) {
-      return nullptr;
+    numInput_ = input_->size();
+    if (replicateNullsAndAny_) {
+      collectNullRows();
     }
-    collectNullRows();
+    // We will be processing the input elements before any other.
+    currentState_ = ProcessingState::kProcessingInputElements;
 
-    const vector_size_t maxOutputSize = calculateMaxOutputSize();
+    velox::row::UnsafeRowFast unsafeRow(reorderInputsIfNeeded());
+    calculateSerializedRowSize(unsafeRow);
 
-    // TODO Reuse output vector.
-    auto output =
-        BaseVector::create<RowVector>(outputType_, maxOutputSize, pool());
-    auto partitionVector = output->childAt(0)->asFlatVector<int32_t>();
-    partitionVector->resize(maxOutputSize);
-    auto dataVector = output->childAt(1)->asFlatVector<StringView>();
-    dataVector->resize(maxOutputSize);
+    outputVector_ = createOutputRowVector(numInput_);
+    auto partitionVector = outputVector_->childAt(0)->asFlatVector<int32_t>();
+    auto dataVector = outputVector_->childAt(1)->asFlatVector<StringView>();
 
     computePartitions(*partitionVector);
-    serializeRows(*dataVector);
-
-    if (replicateNullsAndAny_ && numPartitions_ > 1) {
-      auto numOutputRows =
-          replicateRowsWithNullPartitionKeys(*dataVector, *partitionVector);
-      // If numOutputRows equal to the input row count, then
-      // 'replicateRowsWithNullPartitionKeys' did not insert any row into the
-      // output vector, thus we need to call into
-      // 'replicateAnyToEmptyPartitions'
-      if (numOutputRows == input_->size()) {
-        numOutputRows = replicateAnyToEmptyPartitions(
-            *partitionVector, *dataVector, numOutputRows);
+    serializeRows(unsafeRow, *dataVector);
+    numOutputRows_ = numInput_;
+    input_ = nullptr;
+    if (replicateNullsAndAny_) {
+      nonEmptyPartitions_.insert(
```

Comment:
It looks like there is a good amount of logic and complexity to optimize for not replicating 'any' row to non-empty partitions. Is this the right tradeoff? Consider replicating first row to all partitions.

Reply:
> In addition, there could be edge case that the first batch of input is all null we will then need to wait for the set of input

In this case we don't need to replicate 'any' row at all.